Update federated_cloud_sharing_configuration.adoc (#1087)
owncloud/docs-server: modules/admin_manual/pages/configuration/files/federated_cloud_sharing_configuration.adoc

== Introduction

Federated Cloud Sharing is managed by the Federation app. When you enable the Federation app you can easily and securely link file shares between ownCloud servers, in effect creating a "cloud" of ownCloud installations.

IMPORTANT: For security reasons federated sharing **strictly requires HTTPS (SSL/TLS)**.

IMPORTANT: We strongly recommend using HTTP for development and testing purposes. However, to do so, you have to set `'sharing.federation.allowHttpFallback' => true,` in **config/config.php**.

== Configuration

Follow these steps to establish a trusted connection between two servers.

> . Verify that both servers have SSL certificates. +

Reviewer: Is the "+" supposed to be there?

Reviewer 2: Yes. It tells the (numbered ) list item that the current sentence ends here, but the next line(s) are part of the item and therefore properly intented. (soft break). I can make you a build on staging so you can "optical review" this. Before commiting this change, a final local build rendered fine.

If you open the server URL in your browser and see a lock icon on the left-hand side of the address bar, the certificate is valid:
+
.Lock icon in the address bars in Firefox, Google Chrome, and Safari
image::configuration/files/browser-address-bars.png[Lock icon in Firefox, Google Chrome, and Safari]

. Verify that the `'overwrite.cli.url' => 'https://<SERVER_URL>'` setting is configured to the correct URL instead of `localhost` in *config.php*:

. Navigate to menu:Admin settings[Sharing > Federation]

. Add **server 1** to the trusted servers on **server 2**

. Add **server 2** to the trusted servers on **server 1**

. Run `system:cron` from the ownCloud directory of both servers, for example `/var/www/owncloud/`:
+
[source,bash,subs="attributes+"]
----
{occ-command-example-prefix} system:cron
----

. Run the `OCA\Federation\SyncJob` job with the `force` option on both servers. +
You can get the corresponding job ID by using the `background:queue:status` occ command:
+
[source,bash,subs="attributes+"]
----
{occ-command-example-prefix} background:queue:execute --force <jobId>
----

. The check should now be green

> . Sync now your users by running on both servers

Reviewer: ```suggestion
. Now sync your users by running on both servers
```

Or leave out "now".

+
[source,bash,subs="attributes+"]
----
{occ-command-example-prefix} dav:sync-system-addressbook
{occ-command-example-prefix} federation:sync-addressbook
----

. Configure automatic acceptance of new federated shares
+
--
NOTE: Automatic acceptance of new federated shares will not work if the option `Add server automatically once a federated share was created successfully` is also set. This is done because of security concerns.

[source,bash,subs="attributes+"]
----
{occ-command-example-prefix} config:app:set federation auto_accept_trusted --value '0'
{occ-command-example-prefix} config:app:set federatedfilesharing auto_accept_trusted --value 'yes'
----
--

== Working With Proxies

There are ownCloud instances that are not connected to the internet. They have no possibility to reach the public network. Therefore Federation will not work without a proxy.

> To set the `proxy` and `proxyuserpwd` configuration variables, in `config/config.php`. `proxy` sets the proxy’s hostname, and `proxyuserpwd` sets the username and password credentials, in `username:password` format.

Reviewer: ```suggestion
Set the `proxy` and `proxyuserpwd` configuration variables, in `config/config.php`. `proxy` sets the proxy’s hostname, and `proxyuserpwd` sets the username and password credentials, in `username:password` format.
```

Follow these steps to establish a trusted connection between two servers.

. Verify that both servers have SSL certificates
  If you open the server URL in your browser and see a lock icon on the left-hand side of the address bar, the certificate is valid
+
.Lock icon in the address bars in Firefox, Google Chrome, and Safari.
image:configuration/files/browser-address-bars.png[Lock icon in Firefox, Google Chrome, and Safari]
. Verify that the `'overwrite.cli.url' => 'https://<SERVER_URL>'` setting is configured to the correct URL instead of `localhost` in **config.php**
. Navigate to **admin settings -> Sharing -> Federation**
. Add **server 1** to the trusted servers on **server 2**
. Add **server 2** to the trusted servers on **server 1**

> . Run `system:cron` from the ownCloud directory of both servers (for example `/var/www/owncloud/`)

Reviewer 3: ```suggestion
. Run `{occ-command-example-prefix} system:cron` in both servers
```

Wrong place to discuss if it is needed to really chdir somwhere, when running occ commands. (And it should not be needed to chdir in anycase, imo, I never do)

But occ should be mentioned, when running occ commands.

Reviewer 2: I have further improved the text based on your changes, see commit, pls approve

+
[source,bash,subs="attributes+"]
----
{occ-command-example-prefix} system:cron
----

> . Run the `OCA\Federation\SyncJob` job with the `force` option on both servers (you can get the correspondent job ID by using the `background:queue:status` occ command)

Reviewer: ```suggestion
. Run the `OCA\Federation\SyncJob` job with the `force` option on both servers (you can get the corresponding job ID by using the `background:queue:status` occ command)
```

+
[source,bash,subs="attributes+"]
----

> {occ-command-example-prefix} background:queue:execute --force 10

Reviewer 3: Is that sufficient?
I am not sure, if maybe we still need to call occ system:cron 
t really execute the job?

Author: Based on my tests it seems to be sufficient.

Reviewer 4: we cannot be sure that id 10 will be what we need. it varies per system

Author: correct but there is also documented `you can get the correspondent job ID by using the background:queue:status occ command`.

Reviewer 3: How about something like:
```
fedJobId=$(occ background:queue:status | grep Federation.SyncJob | cut -d '|' -f 2)
occ background:queue:execute --force $fedJobId
```

Reviewer 3: I've checked a few of my systems, it is always 10 there. I really thought that is a constant. :frowning:

Reviewer 3: It never works for me on the first attempt.
I have now succeeded in one case, where I did exactly as described here, 
got only a yellow marker in the webUI, then pondered a while, and re-ran (a few minutes later) 
`occ background:queue:execute --force $fedJobId`
This additional (or delayed) command then did the magic: It added the entry to oc_addressbooks, and the trusted server became green in the UI.

> {occ-command-example-prefix} background:queue:execute --force 10

Reviewer: ```suggestion
{occ-command-example-prefix} background:queue:execute --force <jobId>
```

Can we just mention `<jobId>` something like this?
How do we usually indicate to users that they need to substitute their own value in a command?

Reviewer 3: Also, I'd suggest to not only use `--force` here, (which does not force much, it still scares the user with a fat where he has to type 'yes') but use `--force --accept-warning` which really runs the command without asking questions.

Author: `--force` is rather meant for running the job even if it is already within the time interval. So the time interval will be ignored and `last_run` will be set to `0`. But yes, also using `--accept-warning` makes sense.

Reviewer 2: Mind to adapt @phil-davis suggestable commit with whatever option is correct or commit the suggestion if it is ok? Else this stays open and nobody cares...

=== Persistent Locks Are Not Guaranteed

> There is a known bug propagated persistent locks to federated instances. If a user creates an exclusive lock on a share, no other users should be able to modify it, nor its contents, and all users should see a lock icon on the share.

Reviewer: ```suggestion
There is a known bug in propagation of persistent locks to federated instances. If a user creates an exclusive lock on a share, no other users should be able to modify it, nor its contents, and all users should see a lock icon on the share.
```
